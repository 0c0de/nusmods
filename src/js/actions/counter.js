const INCREMENT = 'INCREMENT';
export function increment() {
  return {
    type: INCREMENT
  };
}

const DECREMENT = 'DECREMENT';
export function decrement() {
  return {
    type: DECREMENT
  };
}
