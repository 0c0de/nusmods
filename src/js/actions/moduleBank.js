import { API_REQUEST } from 'middlewares/requests-middleware';
import NUSModsApi from 'utils/nusmods-api';

export const GET_MODULE_LIST = 'GET_MODULE_LIST';
export function getModuleList() {
  return (dispatch) => dispatch({
    [API_REQUEST]: {
      type: GET_MODULE_LIST,
      payload: {
        method: 'GET',
        url: NUSModsApi.moduleListUrl(),
      },
    },
  });
}

export const GET_MODULE = 'GET_MODULE';
export function getModule(moduleCode) {
  return (dispatch, getState) => {
    // Module has been fetched before and cached. Don't have to fetch again.
    if (getState().entities.moduleBank.modules[moduleCode]) {
      return null;
    }

    return dispatch({
      [API_REQUEST]: {
        type: GET_MODULE,
        payload: {
          method: 'GET',
          url: NUSModsApi.moduleDetailsUrl(moduleCode),
        },
      },
    });
  };
}
