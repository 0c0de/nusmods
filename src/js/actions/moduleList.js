import { API_REQUEST } from 'middlewares/requests-middleware';
import NUSModsApi from 'utils/nusmods-api';

export const GET_MODULE_LIST = 'GET_MODULE_LIST';
export function getModuleList() {
  return (dispatch) => dispatch({
    [API_REQUEST]: {
      type: GET_MODULE_LIST,
      payload: {
        method: 'GET',
        url: NUSModsApi.moduleListUrl(),
      },
    },
  });
}
