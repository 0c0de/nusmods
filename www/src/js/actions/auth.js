// @flow
import { generateAuthActions } from 'redux-token-auth';
import config from 'config';

const rtaConfig = {
  authUrl: `${config.railsApiBaseUrl}/auth`,
};

const { registerUser, signInUser, signOutUser, verifyCredentials } = generateAuthActions(rtaConfig);
export { registerUser, signInUser, signOutUser, verifyCredentials };
