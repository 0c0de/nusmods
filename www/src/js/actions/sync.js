// @flow
import type { FSA } from 'types/redux';

export const SYNC_DATA_RECEIVED = 'SYNC_DATA_RECEIVED ';
export function syncDataReceived(data: Object): FSA {
  return {
    type: SYNC_DATA_RECEIVED,
    payload: { data },
  };
}
