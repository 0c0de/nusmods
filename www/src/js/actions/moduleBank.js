// @flow
import type { ModuleCode } from 'types/modules';
import type { FSA } from 'types/redux';

import { requestAction } from 'actions/requests';
import NUSModsApi from 'apis/nusmods';
import config from 'config';

export const FETCH_MODULE_LIST: string = 'FETCH_MODULE_LIST';
export function fetchModuleList(): FSA {
  return requestAction(FETCH_MODULE_LIST, FETCH_MODULE_LIST, {
    url: NUSModsApi.moduleListUrl(),
  });
}

// Action to fetch module
export const FETCH_MODULE: string = 'FETCH_MODULE';
export function fetchModuleRequest(moduleCode: ModuleCode) {
  return `${FETCH_MODULE}_${moduleCode}`;
}

export function fetchModule(moduleCode: ModuleCode): FSA {
  const key = fetchModuleRequest(moduleCode);
  return requestAction(key, FETCH_MODULE, {
    url: NUSModsApi.moduleDetailsUrl(moduleCode),
  });
}

// Action to fetch module from previous years
export const FETCH_ARCHIVE_MODULE: string = 'FETCH_ARCHIVE_MODULE';
export function fetchArchiveRequest(moduleCode: ModuleCode, year: string) {
  return `${FETCH_ARCHIVE_MODULE}_${moduleCode}_${year}`;
}

export function fetchModuleArchive(moduleCode: ModuleCode, year: string) {
  const key = fetchArchiveRequest(moduleCode, year);
  return requestAction(key, FETCH_ARCHIVE_MODULE, {
    url: NUSModsApi.moduleDetailsUrl(moduleCode, year),
  });
}

export function fetchAllModuleArchive(moduleCode: ModuleCode) {
  return (dispatch: Function) =>
    Promise.all(
      config.archiveYears.map((year) =>
        dispatch(fetchModuleArchive(moduleCode, year)).catch(() => null),
      ),
    ).then((modules) => modules.filter(Boolean));
}
