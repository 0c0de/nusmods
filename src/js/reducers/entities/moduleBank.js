import { GET_MODULE_LIST } from 'actions/moduleBank';
import * as RequestResultCases from 'middlewares/requests-middleware';

const defaultModuleBankState = {
  moduleList: [],    // List of modules
  modules: {},       // Object of ModuleCode -> ModuleDetails
};

function moduleBank(state = defaultModuleBankState, action) {
  switch (action.type) {
    case GET_MODULE_LIST + RequestResultCases.SUCCESS:
      return Object.assign({}, state, {
        moduleList: action.response,
      });
    default:
      return state;
  }
}

export default moduleBank;
