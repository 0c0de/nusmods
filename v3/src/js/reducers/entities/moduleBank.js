// @flow
import type { FSA } from 'redux';
import type {
  Module,
  ModuleCode,
  ModuleCondensed,
  Semester,
} from 'types/modules';
import type {
  SemTimetableConfig,
} from 'types/timetable';

import _ from 'lodash';

import { FETCH_MODULE_LIST, FETCH_MODULE } from 'actions/moduleBank';
import * as RequestResultCases from 'middlewares/requests-middleware';

type ModuleSelectListItem = {
  label: string,
  value: ModuleCode,
  semesters: Array<number>
};
export type ModuleBank = {
  moduleList: Array<ModuleCondensed>,
  moduleSelectList: Array<ModuleSelectListItem>,
  modules: {
    [key: ModuleCode]: Module
  },
};

const defaultModuleBankState: ModuleBank = {
  moduleList: [],     // List of modules
  moduleSelectList: [],
  modules: {},       // Object of ModuleCode -> ModuleDetails
};

function moduleBank(state: ModuleBank = defaultModuleBankState, action: FSA): ModuleBank {
  switch (action.type) {
    case FETCH_MODULE_LIST + RequestResultCases.SUCCESS:
      return {
        ...state,
        moduleList: action.response,
        // Precompute this in reducer because putting this inside render is very expensive (5k modules!)
        moduleSelectList: action.response.map((module: ModuleCondensed): ModuleSelectListItem => {
          return {
            semesters: module.Semesters,
            value: module.ModuleCode,
            label: `${module.ModuleCode} ${module.ModuleTitle}`,
          };
        }),
      };
    case FETCH_MODULE + RequestResultCases.SUCCESS:
      return {
        ...state,
        modules: {
          ...state.modules,
          [action.response.ModuleCode]: action.response,
        },
      };
    default:
      return state;
  }
}

export function getSemModuleSelectList(state: ModuleBank, semester: Semester,
                                        semTimetableConfig: SemTimetableConfig): Array<ModuleSelectListItem> {
  if (!state.moduleSelectList) {
    return [];
  }
  return state.moduleSelectList.filter((item: ModuleSelectListItem): boolean => {
    // In specified semester and not within the timetable.
    return _.includes(item.semesters, semester) && !semTimetableConfig[item.value];
  });
}

export default moduleBank;
