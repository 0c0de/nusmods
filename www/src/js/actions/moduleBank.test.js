// @flow
import * as actions from 'actions/moduleBank';
import NUSModsApi from 'apis/nusmods';

// Mock NUSModsApi as its URLs contain the current AY, breaking the snapshot tests
// every AY.
jest.mock('apis/nusmods');
NUSModsApi.moduleListUrl.mockReturnValue('test://MOCK_MOD_LIST_URL');
NUSModsApi.moduleDetailsUrl.mockImplementation(
  (...args) => `test://MOCK_MOD_DETAILS_URL/${[...args].join('/')}`,
);

test('fetchModuleList should return a request action', () => {
  const resultOfAction = actions.fetchModuleList();
  expect(resultOfAction).toMatchSnapshot();
});

test('fetchModule should return a request action', () => {
  const resultOfAction = actions.fetchModule('CS1010S');
  expect(resultOfAction).toMatchSnapshot();
});

test('fetchModuleArchive should return a request action', () => {
  expect(actions.fetchModuleArchive('CS1010S', '2016/2017')).toMatchSnapshot();
});

test('fetchAllModuleArchive should return multiple request actions', () => {
  const dispatch = jest.fn().mockReturnValue(Promise.resolve());
  const thunk = actions.fetchAllModuleArchive('CS1010S');
  expect(thunk).toEqual(expect.any(Function));
  thunk(dispatch);
  expect(dispatch.mock.calls).toMatchSnapshot();
});
