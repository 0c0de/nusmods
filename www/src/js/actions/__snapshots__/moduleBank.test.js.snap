// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchAllModuleArchive should return multiple request actions 1`] = `
Array [
  Array [
    Object {
      "meta": Object {
        "API_REQUEST": "FETCH_ARCHIVE_MODULE_CS1010S_2016/2017",
        "academicYear": "2016/2017",
      },
      "payload": Object {
        "url": "test://MOCK_MOD_DETAILS_URL/CS1010S/2016/2017",
      },
      "type": "FETCH_ARCHIVE_MODULE",
    },
  ],
  Array [
    Object {
      "meta": Object {
        "API_REQUEST": "FETCH_ARCHIVE_MODULE_CS1010S_2017/2018",
        "academicYear": "2017/2018",
      },
      "payload": Object {
        "url": "test://MOCK_MOD_DETAILS_URL/CS1010S/2017/2018",
      },
      "type": "FETCH_ARCHIVE_MODULE",
    },
  ],
]
`;

exports[`fetchModule should return a request action 1`] = `
Object {
  "meta": Object {
    "API_REQUEST": "FETCH_MODULE_CS1010S",
  },
  "payload": Object {
    "url": "test://MOCK_MOD_DETAILS_URL/CS1010S",
  },
  "type": "FETCH_MODULE",
}
`;

exports[`fetchModuleArchive should return a request action 1`] = `
Object {
  "meta": Object {
    "API_REQUEST": "FETCH_ARCHIVE_MODULE_CS1010S_2016/2017",
    "academicYear": "2016/2017",
  },
  "payload": Object {
    "url": "test://MOCK_MOD_DETAILS_URL/CS1010S/2016/2017",
  },
  "type": "FETCH_ARCHIVE_MODULE",
}
`;

exports[`fetchModuleList should return a request action 1`] = `
Object {
  "meta": Object {
    "API_REQUEST": "FETCH_MODULE_LIST",
  },
  "payload": Object {
    "url": "test://MOCK_MOD_LIST_URL",
  },
  "type": "FETCH_MODULE_LIST",
}
`;
